substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/PosixJavaNetSubstitutions.java

```diff
@@ -119,84 +419,1339 @@ protected int peekData(DatagramPacket p) throws IOException {
+
+        //   276      if (NET_InetAddressToSockaddr(env, address, port, (struct sockaddr *)&rmtaddr, &len, JNI_TRUE) != 0) {
+        if (JavaNetNetUtilMD.NET_InetAddressToSockaddr(address, port, rmtaddr, len_Pointer, true) != 0)  {
+            //   277        return;
+            return;
+        }
+
+        //   280      setDefaultScopeID(env, (struct sockaddr *)&rmtaddr);
+        JavaNetNetUtilMD.setDefaultScopeID(rmtaddr);
+        //   281
+        //   282      if (JVM_Connect(fd, (struct sockaddr *)&rmtaddr, len) == -1) {
+        if (VmPrimsJVM.JVM_Connect(fd, rmtaddr, len_Pointer.read()) == -1) {
+            //   283          NET_ThrowByNameWithLastError(env, JNU_JAVANETPKG "ConnectException",
+            //   284                          "Connect failed");
+            throw new ConnectException(PosixUtils.lastErrorString("Connect failed"));
+            //   285          return;
+        }
     }
+    // @formatter:on
 
 }
 
 @TargetClass(java.net.DatagramSocket.class)
 @Platforms({Platform.LINUX.class, Platform.DARWIN.class})
 final class Target_java_net_DatagramSocket {
 
-    @Substitute
-    @SuppressWarnings("unused")
-    Target_java_net_DatagramSocket(SocketAddress bindaddr) throws SocketException {
-        throw VMError.unsupportedFeature("Unimplemented: java.net.DatagramSocket.<init>(SocketAddress)");
-    }
-
     @Substitute
     private void checkOldImpl() {
         // it calls java.net.PlainDatagramSocketImpl.peekData(DatagramPacket) which is unimplemented
     }
 
-    @Substitute
-    @SuppressWarnings({"static-method", "unused"})
-    public void bind(SocketAddress addr) throws SocketException {
-        throw VMError.unsupportedFeature("Unimplemented: java.net.DatagramSocket.bind(SocketAddress)");
+}
+
+@TargetClass(java.net.DatagramPacket.class)
+@Platforms({Platform.LINUX.class, Platform.DARWIN.class})
+final class Target_java_net_DatagramPacket {
+    /* Aliases to get visibility for substituted methods. */
+    @Alias byte[] buf;
+    @Alias InetAddress address;
+    @Alias int offset;
+    @Alias int length;
+    @Alias int bufLength;
+    @Alias int port;
+}
+
+class Util_java_net_DatagramPacket {
+    static Target_java_net_DatagramPacket as_Target_java_net_DatagramPacket(DatagramPacket packet) {
+        return KnownIntrinsics.unsafeCast(packet, Target_java_net_DatagramPacket.class);
     }
 }
 
+@TargetClass(java.net.DatagramSocketImpl.class)
+@Platforms({Platform.LINUX.class, Platform.DARWIN.class})
+final class Target_java_net_DatagramSocketImpl {
+    /* Aliases to get visibility for substituted methods. */
+    @Alias FileDescriptor fd;
+    @Alias int localPort;
+}
+
+class Util_java_net_DatagramSocketImpl {
+    static Target_java_net_DatagramSocketImpl as_Target_java_net_DatagramSocketImpl(Target_java_net_PlainDatagramSocketImpl tjnsi) {
+        return KnownIntrinsics.unsafeCast(tjnsi, Target_java_net_DatagramSocketImpl.class);
+    }
+}
+
+class Util_java_net_PlainDatagramSocketImpl {
```

Comment:
This is interesting.  We conventionally use "Util_" classes as collections of new Java methods that are useful to "Target_" implementations.  Here you are using it also as a collection of C "file static" methods from PlainDatagramSocketImpl.c that do not otherwise have a place to live.  I guess that's cool.